### PR TITLE
fix(harness): avoid shell-based sandbox edit

### PR DESCRIPTION
**Pull request by guslegend0510**
## 变更说明

本 PR 修复了 #1609，移除了 `BaseSandboxFilesystem.edit()` 中基于 shell/Python 的编辑实现。

旧实现会拼接 `python3 -c` 命令，并用字面量 `\n` 连接 Python 代码。在 Aone、E2B 等远程沙箱中，这种命令可能被 Python 当成单行代码处理，进而触发 `SyntaxError`。

现在 `edit()` 改为更稳定的流程：

1. 通过沙箱文件接口下载目标文件
2. 在 Java 侧完成字符串替换
3. 将修改后的内容重新上传回沙箱

## 具体改动

- 将 `BaseSandboxFilesystem.edit()` 从 shell/Python 执行改为 download -> replace -> upload
- 保持现有编辑语义不变，包括：
  - 文件不存在
  - 目标字符串不存在
  - `replaceAll=false` 时出现多处匹配
  - 上传/下载失败
  - 文件内容不是合法 UTF-8
- 新增回归测试，确保 `edit()` 不再依赖 `execute()`

## 附带修复

本 PR 同时修复了 `ToolCallParam.builder(source)` 的拷贝构造问题。

当前 `upstream/main` 中，这里错误地使用了 `source.context`，但 Builder 实际字段是 `runtimeContext`，会导致 `agentscope-core` 编译失败。本 PR 一并修正了这处赋值，并补充了相关测试覆盖。

## 验证

执行命令：

```bash
mvn --% -pl agentscope-harness -am -Dtest=ToolCallParamTest,BaseSandboxFilesystemTest,CompositeFilesystemTest,FilesystemDeleteMoveExistsTest,FilesystemGlobTest,LocalFilesystemModeTest,RemoteFilesystemCASTest,RemoteFilesystemGrepTest,RemoteFilesystemSpecTest -Dspotless.check.skip=true test

### CODE REVIEW

File: agentscope-harness/src/main/java/io/agentscope/harness/agent/filesystem/sandbox/BaseSandboxFilesystem.java [hunk 1]
```diff
@@ -54,6 +59,8 @@
  */
 public abstract class BaseSandboxFilesystem implements AbstractSandboxFilesystem {
 
+    private final ConcurrentHashMap<String, ReentrantLock> editLocks = new ConcurrentHashMap<>();
```
Comment (AgentScopeJavaBot) [inline agentscope-harness/src/main/java/io/agentscope/harness/agent/filesystem/sandbox/BaseSandboxFilesystem.java:62]: [minor] The `editLocks` map grows without bound -- every unique `filePath` passed to `edit()` inserts a `ReentrantLock` that is never removed. For a long-lived `BaseSandboxFilesystem` instance that edits many distinct paths across sessions, stale entries accumulate. Consider periodically pruning idle locks (e.g., wrap in a small holder that tracks last-access time and sweep lazily), or accept the trade-off with a brief comment explaining why unbounded growth is acceptable for the expected workload.

File: agentscope-harness/src/main/java/io/agentscope/harness/agent/filesystem/sandbox/BaseSandboxFilesystem.java [hunk 2]
```diff
@@ -379,14 +368,22 @@ public boolean exists(RuntimeContext runtimeContext, String path) {
...
         }
-        return s.replace("\\", "\\\\")
-                .replace("\"", "\\\"")
-                .replace("\n", "\\n")
-                .replace("\r", "\\r")
-                .replace("\t", "\\t");
+        String normalized = error.toLowerCase();
+        return normalized.contains("no such file")
+                || normalized.contains("not found")
+                || normalized.contains("file_not_found");
```
Comment (AgentScopeJavaBot) [inline agentscope-harness/src/main/java/io/agentscope/harness/agent/filesystem/sandbox/BaseSandboxFilesystem.java:387]: [minor] The `"not found"` substring check in `isMissingFileError` is fairly broad and could match non-file-related backend errors (e.g., `"Sandbox not found"`, `"Token not found"`). Since this method is only called on download errors, false positives are unlikely in practice, but consider tightening the heuristic or documenting the assumption.
```java
// Example: more specific patterns
return normalized.contains("no such file")
        || normalized.contains("file not found")
        || normalized.contains("file_not_found");
```

File: agentscope-harness/src/main/java/io/agentscope/harness/agent/filesystem/sandbox/BaseSandboxFilesystem.java [hunk 3]
```diff
@@ -379,14 +368,22 @@ public boolean exists(RuntimeContext runtimeContext, String path) {
         return result.output() != null && result.output().strip().startsWith("yes");
     }
 
-    private static String jsonEscape(String s) {
-        if (s == null) {
-            return "";
+    private String decodeUtf8Strict(byte[] content) throws CharacterCodingException {
```
Comment (AgentScopeJavaBot) [inline agentscope-harness/src/main/java/io/agentscope/harness/agent/filesystem/sandbox/BaseSandboxFilesystem.java:371]: [nit] `decodeUtf8Strict` does not reference any instance state; it could be declared `private static` for clarity.

File: agentscope-harness/src/test/java/io/agentscope/harness/agent/filesystem/sandbox/BaseSandboxFilesystemTest.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        InMemorySandboxFilesystem fs = new InMemorySandboxFilesystem();
+
+        EditResult result = fs.edit(RT, "/workspace/missing.txt", "old", "new", false);
+
+        assertFalse(result.isSuccess());
+        assertEquals("Error: File '/workspace/missing.txt' not found", result.error());
+        assertEquals(1, fs.downloadCalls.get());
+        assertEquals(0, fs.uploadCalls.get());
+        assertEquals(0, fs.executeCalls.get());
+    }
```
Comment (AgentScopeJavaBot) [inline agentscope-harness/src/test/java/io/agentscope/harness/agent/filesystem/sandbox/BaseSandboxFilesystemTest.java:66]: [minor] The two tests cover the happy path and file-not-found, but several edge cases listed in the PR description lack dedicated coverage: `replaceAll=true` with multiple occurrences, string not found in content, multiple occurrences with `replaceAll=false` (error path), invalid UTF-8 file content, and upload failure. Adding at least the `replaceAll` and string-not-found cases would strengthen regression confidence for the new download-replace-upload flow.

File: agentscope-harness/src/main/java/io/agentscope/harness/agent/filesystem/sandbox/BaseSandboxFilesystem.java [hunk 5]
```diff
@@ -192,84 +199,66 @@ public EditResult edit(
...
-                    + "    sys.exit(0)\\n"
-                    + "result = text.replace(old, new) if replace_all else text.replace(old, new,"
-                    + " 1)\\n"
-                    + "with open(path, 'wb') as f: f.write(result.encode('utf-8'))\\n"
-                    + "print(json.dumps({'count': count}))\\n"
-                    + "\" 2>&1 <<'__EDIT_EOF__'\n"
-                        + payloadB64
-                        + "\n__EDIT_EOF__\n";
+        ReentrantLock lock = editLocks.computeIfAbsent(filePath, key -> new ReentrantLock());
+        lock.lock();
```
Comment (AgentScopeJavaBot) [inline agentscope-harness/src/main/java/io/agentscope/harness/agent/filesystem/sandbox/BaseSandboxFilesystem.java:203]: [praise] Good use of per-file `ReentrantLock` to prevent TOCTOU races between the download and upload steps. This ensures concurrent `edit()` calls on the same path are serialized correctly.

File: agentscope-core/src/main/java/io/agentscope/core/tool/ToolCallParam.java [hunk 6]
```diff
@@ -171,7 +171,7 @@ private Builder(ToolCallParam source) {
             this.toolUseBlock = source.toolUseBlock;
             this.input = source.input.isEmpty() ? null : source.input;
             this.agent = source.agent;
-            this.context = source.context;
+            this.runtimeContext = source.runtimeContext;
```
Comment (AgentScopeJavaBot) [inline agentscope-core/src/main/java/io/agentscope/core/tool/ToolCallParam.java:174]: [praise] Correct fix -- `this.context = source.context` referenced non-existent builder field and source field, which would fail to compile. Using `this.runtimeContext = source.runtimeContext` aligns with the actual `Builder` field name and the `ToolCallParam` constructor.

**Review by AgentScopeJavaBot — APPROVED** [review]
🤖 AI Review

This PR replaces a fragile shell/Python-based `edit()` implementation in `BaseSandboxFilesystem` with a robust download-replace-upload flow performed entirely in Java. The old approach assembled a `python3 -c` command with literal `\n` connectors, which triggered `SyntaxError` on remote sandboxes like Aone and E2B. The new approach delegates to existing `downloadFiles`/`uploadFiles` abstractions and the already-proven `FilesystemUtils.performStringReplacement()`, which is a clear improvement in reliability and maintainability. The accompanying fix to `ToolCallParam.builder(source)` corrects a real compilation error where `this.context` was used instead of `this.runtimeContext`. Both changes are well-motivated, correctly implemented, and include relevant test coverage. The per-file `ReentrantLock` addition is a thoughtful concurrency safeguard for the read-modify-write cycle.
